build-system/tasks/build-bento.js

```diff
@@ -0,0 +1,271 @@
+ * @param {string} name Name of the extension. Must be the sub directory in
+ *     the components directory and the name of the JS and optional CSS file.
+ * @param {string} version Version of the extension. Must be identical to
+ *     the sub directory inside the extension directory
+ * @param {string} latestVersion Latest version of the extension.
+ * @param {boolean} hasCss Whether there is a CSS file for this extension.
+ * @param {?Object} options
+ * @param {!Array=} extraGlobs
+ * @return {!Promise<void>}
+ */
+async function buildComponent(
+  name,
+  version,
+  latestVersion,
+  hasCss,
+  options,
+  extraGlobs
+) {
+  options = options || {};
+  options.extraGlobs = extraGlobs;
+  if (options.compileOnlyCss && !hasCss) {
+    return;
+  }
+  const componentsDir = `src/bento/components/${name}/${version}`;
+  if (options.watch) {
+    await watchComponent(
+      componentsDir,
+      name,
+      version,
+      latestVersion,
+      hasCss,
+      options
+    );
+  }
+
+  if (hasCss) {
+    if (!existsSync('build')) {
+      mkdirSync('build');
+    }
+    if (!existsSync('build/css')) {
+      mkdirSync('build/css');
+    }
+    await buildExtensionCss(componentsDir, name, version, options);
+    if (options.compileOnlyCss) {
+      return;
+    }
+  }
+  if (options.npm) {
+    await buildNpmBinaries(componentsDir, name, options);
+    await buildNpmCss(componentsDir, options);
+  }
+  if (options.binaries) {
+    await buildBinaries(componentsDir, options.binaries, options);
+  }
+  if (options.isRebuild) {
+    return;
+  }
+
+  const bentoName = getBentoName(name);
+  await buildExtensionJs(componentsDir, bentoName, {
+    ...options,
+    wrapper: 'none',
+    filename: await getBentoBuildFilename(
+      componentsDir,
+      bentoName,
+      'standalone',
+      options
+    ),
+    // Include extension directory since our entrypoint may be elsewhere.
+    extraGlobs: [...(options.extraGlobs || []), `${componentsDir}/**/*.js`],
+  });
+}
+
+/**
+ * Builds a single component after extracting its settings.
+ * @param {!Object} components
+ * @param {string} component
+ * @param {!Object} options
+ * @return {!Promise<void>}
+ */
+async function doBuildComponent(components, component, options) {
+  const e = components[component];
+  const o = {...options, ...e};
+  await buildComponent(
+    e.name,
+    e.version,
+    e.latestVersion,
+    e.hasCss,
+    o,
+    e.extraGlobs
+  );
+}
+
+/**
+ * Build all the Bento components
+ *
+ * @param {!Object} options
+ * @return {!Promise<void>}
+ */
+async function buildBentoComponents(options) {
+  const startTime = Date.now();
+  maybeInitializeComponents();
+  const toBuild = getComponentsToBuild();
+  const results = [];
+  for (const component in components) {
+    if (
+      options.compileOnlyCss ||
+      toBuild.includes(components[component].name)
+    ) {
+      results.push(doBuildComponent(components, component, options));
+    }
+  }
+  await Promise.all(results);
```

Comment:
It feels odd to me that `doBuildComponent` takes an object and a key, then immediately does `components[component]` with it; likewise, we're looping over keys here, then indexing. If we instead loop over the values directly and pass them directly, with `doBuildComponent accepting the component object directly, I think it becomes easier to follow.

Feel free to ignore the `map`/functional approach, but either way it seems like looping over/passing the component objects directly could be cleaner here

```suggestion
  await Promise.all(
    Object.values(components)
      .filter(c => options.compileOnlyCss || toBuild.includes(c.name))
      .map(c => doBuildComponent(c, options))
  );
```